### PR TITLE
Use new job search API for triggering Databricks job by name

### PR DESCRIPTION
**Pull request by alexott**
Existing implementation uses list API to find all possible jobs and filtering out them by name.  This leads to huge load on the API backend for workspaces with big number of jobs (> 1k).  Databricks Workflows team extended list API with search functionality.

### CODE REVIEW

**Comment by alexott** [comment]
Instead of explicit parameter we can check for presence of an environment variable. Not sure what is the best way, especially when we depreciate old implementation completely

**Comment by o-nikolas** [comment]
> Instead of explicit parameter we can check for presence of an environment variable. Not sure what is the best way, especially when we depreciate old implementation completely

Could you not attempt to make the new API request and then handle the error/exception you get back if it isn't supported by the backend being used, then automatically fall back to the old api in that case?

**Comment by alexott** [comment]
@o-nikolas it's not really a completely new API, it's a new parameter to that API. If API doesn't support that parameter it just returns all data as before...

**Comment by o-nikolas** [comment]
> @o-nikolas it's not really a completely new API, it's a new parameter to that API. If API doesn't support that parameter it just returns all data as before...

If that's the case, then it should be easy to build this in a backwards compatible way without the `use_old_list_impl` flag right?

File: airflow/providers/databricks/hooks/databricks.py [hunk 1]
```diff
@@ -171,22 +171,41 @@ def list_jobs(self, limit: int = 25, offset: int = 0, expand_tasks: bool = False
...
         """
         Finds job id by its name. If there are multiple jobs with the same name, raises AirflowException.
 
         :param job_name: The name of the job to look up.
+        :param use_old_list_impl: if we should use old implementation by using list all API.
         :return: The job_id as an int or None if no job was found.
         """
-        all_jobs = self.list_jobs()
-        matching_jobs = [j for j in all_jobs if j["settings"]["name"] == job_name]
+        if use_old_list_impl:  # For compatibility with old implementation
```
Comment (ephraimbuddy) [inline airflow/providers/databricks/hooks/databricks.py]: I wonder if it should be to `use_new_list_impl` unless this is for a major version bump

Comment (alexott) [reply airflow/providers/databricks/hooks/databricks.py]: I'll reimplement it a different way - that's why I converted it to draft.

**Comment by alexott** [comment]
I've reimplemented it differently, so old review comments aren't much relevant

File: airflow/providers/databricks/hooks/databricks.py [hunk 2]
```diff
@@ -152,29 +152,38 @@ def submit_run(self, json: dict) -> int:
...
         :param expand_tasks: Whether to include task and cluster details in the response.
+        :param job_name: Optional name of a job to search.
         :return: A list of jobs.
         """
         has_more = True
         jobs = []
 
         while has_more:
-            json = {
+            jsn = {
```
Comment (ephraimbuddy) [inline airflow/providers/databricks/hooks/databricks.py]: How about we call this `json_data`? I feel it's better than an abbreviation. No strong opinion here

File: airflow/providers/databricks/hooks/databricks.py [hunk 3]
```diff
@@ -152,29 +152,38 @@ def submit_run(self, json: dict) -> int:
...
-                "offset": offset,
                 "expand_tasks": expand_tasks,
+                "offset": offset,
             }
-            response = self._do_api_call(LIST_JOBS_ENDPOINT, json)
-            jobs += response["jobs"] if "jobs" in response else []
+            if job_name:
+                jsn["name"] = job_name
+            response = self._do_api_call(LIST_JOBS_ENDPOINT, jsn)
+            jbs = response.get("jobs", [])
```
Comment (ephraimbuddy) [inline airflow/providers/databricks/hooks/databricks.py]: These abbreviations feel somehow, `databrick_jobs`...

**Comment by alexott** [comment]
@ephraimbuddy comments are addressed & `mypy` error has been fixed. Can you please review again?
thank you

**Review by potiuk — APPROVED** [review]

**Pull request merged**